### ISSUE TITLE
[Fix] Losses scale with sequence length instead of averaging per token

### ISSUE THREAD

**Issue by elouanzer** [elouanzer/shrinkai#10]
## Motivation

Flagged in community feedback on the announcement thread (r/learnmachinelearning): `nn.KLDivLoss(reduction="batchmean")` divides the summed KL by `input.size(0)` only, i.e. the batch size `B`. It is correct when the input have dimension `[B, C]` (one distribution per example), but it is not for anything with an extra dimension, since PyTorch has no notion of "average over positions" built into `batchmean` (it just sums everything else).

Three losses call `KLDivLoss`/`F.kl_div(reduction="batchmean")` directly on tensors thathave such an extra dimension, without ever reshaping first:

| Loss | Shape actually used | What gets summed (not averaged) before dividing by `B` |
|---|---|---|
| `ReverseKLLoss` ([logits.py](https://github.com/elouanzer/shrinkai/blob/main/src/shrinkai/distillation/losses/logits.py)) | `[B, S, C]` (documented, LLM use case) | sequence positions `S` |
| `JSDLoss` (same file) | `[B, S, C]` | sequence positions `S` |
| `AttentionMapLoss(loss_type="kl")` ([features.py](https://github.com/elouanzer/shrinkai/blob/main/src/shrinkai/distillation/losses/features.py)) | `[B, H, S, S]` (attention maps) | heads `H` **and** query positions `S` |

Verified directly by reading each `forward()`: none of the three reshape before calling `kl_div`. `ReverseKLLoss`'s own docstring lists `[B, C]` and `[B, S, C]` as equally supported shapes without ever noting that the numerical meaning of the loss differs between the two.

## Concrete impact

In the tutorial 03, `HybridLoss(logit_loss=ReverseKLLoss(temperature=4.0), feature_loss=attention_loss, feature_weight=10.0)` combines `ReverseKLLoss` (the S-dependent quantity above) with `AttentionMapLoss(loss_type="mse")` (`F.mse_loss`, a true size-invariant mean). `feature_weight=10.0` is implicitly calibrated against `ReverseKLLoss`'s inflation at this tutorial's specific `max_length=128`, change the sequence length and that balance shifts for reasons that have nothing to do with either model's quality.

Separately: none of the three accept a padding mask, so even after fixing the length-scaling, padded positions (this tutorial uses `pad_token = eos_token`) still get summed into the "average" as if they were real tokens, in any batch with variable-length real content.

## Proposed fix

Flatten every leading dimension except the last (class/vocab) one into a single dimension before calling `kl_div`, so `batchmean` divides by the true number of "rows" being compared instead of just `B`:

- `ReverseKLLoss` / `JSDLoss`: `[B, S, C] -> [B*S, C]` (no-op for already-2D `[B, C]` inputs  (`B*1 == B`) so this is backward-compatible for the classification use case).
- `AttentionMapLoss("kl")`: `[B, H, S, S] -> [B*H*S, S]`.

Concretely, something like a shared helper (mirroring the existing `_apply_feature_distance` pattern in `base.py`, or a variant of the `expects_logits` decorator in `logits.py`):

```python
def _flatten_for_batchmean(x: torch.Tensor) -> torch.Tensor:
    """Collapses every dim but the last into one, so KLDivLoss's
    reduction="batchmean" divides by the true number of compared
    distributions instead of just the batch size."""
    return x.reshape(-1, x.shape[-1])
```

used as `self.kl_div(input=_flatten_for_batchmean(x), target=_flatten_for_batchmean(y))` in all three call sites.

This fixes the length-dependent scaling. It does not by itself fix the padding issue that needs an optional mask threaded through, which is a
bigger change (see below), so worth deciding whether to bundle both or ship the reshape fix first.

## Equations

Current (all three, generic form over some "extra" dimension of size $N$:
$S$ for the logit losses, $H \cdot S$ for attention maps):

$$
L_{\text{current}} = \frac{1}{B} \sum_{b=1}^{B} \sum_{n=1}^{N} \text{KL}\big(p_{b,n} \parallel q_{b,n}\big)
$$

grows linearly with $N$ for a constant per-position fit quality.

Proposed (true per-token/per-position average):

$$
L_{\text{fixed}} = \frac{1}{B \cdot N} \sum_{b=1}^{B} \sum_{n=1}^{N} \text{KL}\big(p_{b,n} \parallel q_{b,n}\big)
$$

## Open questions

- Padding: should these losses gain an optional `attention_mask` / `padding_mask` argument, excluding masked positions from both the sum and the denominator? If so, does that thread through `HybridLoss` and `Distiller.fit()` cleanly, or does it stay a "call the loss directly, not through `Distiller`" advanced use case for now?
- Should the shared flattening helper live in `base.py` (available to any future loss) or stay local to `logits.py`/`features.py`?
- `feature_weight` in the tutorial 03 notebook was tuned against the *current* (buggy) magnitude — fixing this changes the effective loss scale, so the tutorial's example weights and commentary need re-checking once this lands, not just the library code.

## References

- `torch.nn.KLDivLoss` docs (`reduction='batchmean'` semantics):
  https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html
